File: src/main/java/cc/duduhuo/git/conflict/setting/ColorSettingsPanelImpl.java
package cc.duduhuo.git.conflict.setting;

import cc.duduhuo.git.conflict.*;
import cc.duduhuo.git.conflict.tool.BundleTools;
import cc.duduhuo.git.conflict.action.HighlightConflictAction;
import cc.duduhuo.git.conflict.model.MarkColor;
import cc.duduhuo.git.conflict.model.PersistentState;
import cc.duduhuo.git.conflict.tool.Tools;
import com.intellij.ui.JBColor;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.util.*;
import java.util.List;

/**
 * =======================================================
 * Author: liying - liruoer2008@yeah.net
 * Datetime: 2018/4/22 16:50
 * Description:
 * Remarks:
 * =======================================================
 */
public class ColorSettingsPanelImpl extends ColorSettingsPanel {
    private MarkColor mOldMarkColor = new MarkColor();
    private LinkedHashMap<String, MarkColor> mMarkColors;
    private String mSchemeName;
    private List<String> mAllSchemeNames = new ArrayList<>();

    public ColorSettingsPanelImpl() {
        initUI();
        setListeners();
    }

    private void setListeners() {
        cCurrentColor.addMouseListener(new ColorChooserListener(ColorChooserListener.CURRENT));
        cCurrentTitleColor.addMouseListener(new ColorChooserListener(ColorChooserListener.CURRENT_TITLE));
        cIncomingColor.addMouseListener(new ColorChooserListener(ColorChooserListener.INCOMING));
        cIncomingTitleColor.addMouseListener(new ColorChooserListener(ColorChooserListener.INCOMING_TITLE));

        cbColorScheme.addItemListener(e -> {
            if (e.getStateChange() == ItemEvent.SELECTED) {
                String selectedItem = (String) e.getItem();
                updateUI(mMarkColors.get(selectedItem));
            }
        });
        btnNew.addActionListener(e -> {
            String newSchemeName = JOptionPane.showInputDialog(mainPanel, "Please input new scheme name: ");
            if (newSchemeName == null) {
                return;
            }
            if ("".equals(newSchemeName)) {
                JOptionPane.showMessageDialog(mainPanel, "Scheme name can not be empty!", "Failed", JOptionPane.ERROR_MESSAGE);
                return;
            }
            if (mAllSchemeNames.contains(newSchemeName)) {
                JOptionPane.showMessageDialog(mainPanel, "Scheme name already exists!", "Failed", JOptionPane.ERROR_MESSAGE);
                return;
            }
            lbCurrentContent.setText("0x000000");
            lbCurrentTitle.setText("0x000000");
            lbIncomingContent.setText("0xffffff");
            lbIncomingTitle.setText("0xffffff");
            cCurrentColor.setBackground(JBColor.BLACK);
            cCurrentTitleColor.setBackground(JBColor.BLACK);
            cIncomingColor.setBackground(JBColor.WHITE);
            cIncomingTitleColor.setBackground(JBColor.WHITE);
            mAllSchemeNames.add(newSchemeName);
            cbColorScheme.addItem(newSchemeName);
            cbColorScheme.setSelectedItem(newSchemeName);
            MarkColor markColor = new MarkColor(newSchemeName, false, 0x000000,
                0x000000, 0xffffff, 0xffffff);
            // add to markColor map
            mMarkColors.put(newSchemeName, markColor);
            updateUI(markColor);
        });
        btnDelete.addActionListener(e -> {
            String selectedItem = (String) cbColorScheme.getSelectedItem();
            int confirm = JOptionPane.showConfirmDialog(mainPanel, "Delete scheme: " + selectedItem + " ?", "Delete", JOptionPane.OK_CANCEL_OPTION);
            if (confirm == JOptionPane.OK_OPTION) {
                mMarkColors.remove(selectedItem);
                cbColorScheme.removeItem(selectedItem);
                mAllSchemeNames.remove(selectedItem);
                updateUI(mMarkColors.get((String) cbColorScheme.getSelectedItem()));
            }
        });
    }

    private void initUI() {
        PersistentState persistentState = GlobalSettings.getPersistentState();
        mMarkColors = persistentState.getMarkColors();
        mSchemeName = persistentState.getSchemeName();
        String name;
        // make sure the built-in color is displayed at the top.
        cbColorScheme.addItem(BuiltInColor.AUTO_SCHEME_NAME);
        cbColorScheme.addItem(BuiltInColor.INTELLIJ_SCHEME_NAME);
        cbColorScheme.addItem(BuiltInColor.DARCULA_SCHEME_NAME);
        Set<String> keySet = mMarkColors.keySet();
        for (String key : keySet) {
            name = mMarkColors.get(key).getSchemeName();
            if (!name.equals(BuiltInColor.AUTO_SCHEME_NAME) &&
                !name.equals(BuiltInColor.INTELLIJ_SCHEME_NAME) &&
                !name.equals(BuiltInColor.DARCULA_SCHEME_NAME)) {
                cbColorScheme.addItem(name);
            }
            mAllSchemeNames.add(name);
        }
        cbColorScheme.setSelectedItem(mSchemeName);
        mOldMarkColor.copy(mMarkColors.get(mSchemeName));
        updateUI(mOldMarkColor);
    }

    private void updateUI(MarkColor color) {
        if (color == null) {
            panelCustom.setVisible(false);
            btnDelete.setEnabled(false);
            return;
        }
        if (color.isBuiltIn()) {
            panelCustom.setVisible(false);
            btnDelete.setEnabled(false);
        } else {
            panelCustom.setVisible(true);
            btnDelete.setEnabled(true);
        }
        lbCurrentContent.setText(Tools.int2HexString(color.getCurrentColor()));
        lbCurrentTitle.setText(Tools.int2HexString(color.getCurrentTitleColor()));
        lbIncomingContent.setText(Tools.int2HexString(color.getIncomingColor()));
        lbIncomingTitle.setText(Tools.int2HexString(color.getIncomingTitleColor()));
        cCurrentColor.setBackground(new JBColor(color.getCurrentColor(), color.getCurrentColor()));
        cCurrentTitleColor.setBackground(new JBColor(color.getCurrentTitleColor(), color.getCurrentTitleColor()));
        cIncomingColor.setBackground(new JBColor(color.getIncomingColor(), color.getIncomingColor()));
        cIncomingTitleColor.setBackground(new JBColor(color.getIncomingTitleColor(), color.getIncomingTitleColor()));
    }

    private class ColorChooserListener implements MouseListener {
        static final int CURRENT = 0;
        static final int CURRENT_TITLE = 1;
        static final int INCOMING = 2;
        static final int INCOMING_TITLE = 3;
        private int mFieldIndex;


        public ColorChooserListener(int index) {
            mFieldIndex = index;
        }

        @Override
        public void mouseClicked(MouseEvent e) {
            String selectedItem = (String) cbColorScheme.getSelectedItem();
            MarkColor tempColor = mMarkColors.get(selectedItem);
            Color initialColor = null;
            if (mFieldIndex == CURRENT) {
                initialColor = new JBColor(tempColor.getCurrentColor(), tempColor.getCurrentColor());
            } else if (mFieldIndex == CURRENT_TITLE) {
                initialColor = new JBColor(tempColor.getCurrentTitleColor(), tempColor.getCurrentTitleColor());
            } else if (mFieldIndex == INCOMING) {
                initialColor = new JBColor(tempColor.getIncomingColor(), tempColor.getIncomingColor());
            } else if (mFieldIndex == INCOMING_TITLE) {
                initialColor = new JBColor(tempColor.getIncomingTitleColor(), tempColor.getIncomingTitleColor());
            }

            Color chooseColor = JColorChooser.showDialog(mainPanel, "Choose a color", initialColor);
            if (chooseColor != null) {
                int rgb = chooseColor.getRGB();
                String colorStr = Tools.color2HexString(chooseColor);
                if (mFieldIndex == CURRENT) {
                    cCurrentColor.setBackground(new JBColor(rgb, rgb));
                    lbCurrentContent.setText(colorStr);
                    tempColor.setCurrentColor(Integer.parseInt(colorStr.substring(2), 16));
                    tempColor.setDarkCurrentColor(Integer.parseInt(colorStr.substring(2), 16));
                } else if (mFieldIndex == CURRENT_TITLE) {
                    cCurrentTitleColor.setBackground(new JBColor(rgb, rgb));
                    lbCurrentTitle.setText(colorStr);
                    tempColor.setCurrentTitleColor(Integer.parseInt(colorStr.substring(2), 16));
                    tempColor.setDarkCurrentTitleColor(Integer.parseInt(colorStr.substring(2), 16));
                } else if (mFieldIndex == INCOMING) {
                    cIncomingColor.setBackground(new JBColor(rgb, rgb));
                    lbIncomingContent.setText(colorStr);
                    tempColor.setIncomingColor(Integer.parseInt(colorStr.substring(2), 16));
                    tempColor.setDarkIncomingColor(Integer.parseInt(colorStr.substring(2), 16));
                } else if (mFieldIndex == INCOMING_TITLE) {
                    cIncomingTitleColor.setBackground(new JBColor(rgb, rgb));
                    lbIncomingTitle.setText(colorStr);
                    tempColor.setIncomingTitleColor(Integer.parseInt(colorStr.substring(2), 16));
                    tempColor.setDarkIncomingTitleColor(Integer.parseInt(colorStr.substring(2), 16));
                }
                mMarkColors.put(selectedItem, tempColor);
            }
        }

        @Override
        public void mousePressed(MouseEvent e) {

        }

        @Override
        public void mouseReleased(MouseEvent e) {

        }

        @Override
        public void mouseEntered(MouseEvent e) {

        }

        @Override
        public void mouseExited(MouseEvent e) {

        }
    }

    @NotNull
    @Override
    public String getId() {
        return BundleTools.getValue(Constants.BundleKey.SETTINGS_TITLE);
    }

    @Nls
    @Override
    public String getDisplayName() {
        return this.getId();
    }

    @Nullable
    @Override
    public JComponent createComponent() {
        return mainPanel;
    }

    @Override
    public boolean isModified() {
        if (cbColorScheme.getSelectedItem() != mSchemeName) {
            return true;
        } else {
            return !mOldMarkColor.equals(mMarkColors.get((String) cbColorScheme.getSelectedItem()));
        }
    }

    @Override
    public void reset() {
    }

    @Override
    public void apply() {
        boolean isModified = isModified();
        GlobalSettings.getPersistentState().setMarkColors(mMarkColors);
        GlobalSettings.getPersistentState().setSchemeName((String) cbColorScheme.getSelectedItem());
        if (isModified) {
            mSchemeName = (String) cbColorScheme.getSelectedItem();
            mOldMarkColor.copy(mMarkColors.get(mSchemeName));
            Global.sCurrentColor = mOldMarkColor;
            TextAttr.loadTextAttr();
            HighlightConflictAction.refreshHighlight();
        }
    }
}

File: src/main/java/cc/duduhuo/git/conflict/setting/ColorSettingsPanelImpl.kt
package cc.duduhuo.git.conflict.setting

import cc.duduhuo.git.conflict.BuiltInColor
import cc.duduhuo.git.conflict.Constants
import cc.duduhuo.git.conflict.Global
import cc.duduhuo.git.conflict.GlobalSettings
import cc.duduhuo.git.conflict.action.HighlightConflictAction
import cc.duduhuo.git.conflict.model.MarkColor
import cc.duduhuo.git.conflict.model.PersistentMarkColor
import cc.duduhuo.git.conflict.tool.ext.formatHexA
import com.intellij.ui.ColorChooserService
import java.awt.Color
import java.awt.event.ActionEvent
import java.awt.event.ItemEvent
import java.awt.event.MouseEvent
import java.awt.event.MouseListener
import javax.swing.JComponent
import javax.swing.JOptionPane

/**
 * =======================================================
 * Author: liying - liruoer2008@yeah.net
 * Datetime: 2018/4/22 16:50
 * Description: Settings panel's implementation
 * Remarks:
 * =======================================================
 */
class ColorSettingsPanelImpl : ColorSettingsPanel() {
    companion object {
        const val CURRENT_CONTENT = 0
        const val CURRENT_HEADER = 1
        const val INCOMING_CONTENT = 2
        const val INCOMING_HEADER = 3
        const val COMMON_CONTENT = 4
        const val COMMON_HEADER = 5
    }

    private val markColorInUse = MarkColor()
    private lateinit var markColorMap: LinkedHashMap<String, MarkColor>
    private lateinit var schemeInUse: String
    private var autoDetectConflictsWhenFileOpened: Boolean = false
    private val allSchemeNames: MutableList<String> = mutableListOf()

    init {
        initUI()
        setListeners()
    }

    private fun convertPersistentMarkColorMapToMarkColorMap(persistentMarkColorMap: LinkedHashMap<String, PersistentMarkColor>): LinkedHashMap<String, MarkColor> {
        val markColorMap = LinkedHashMap<String, MarkColor>()
        persistentMarkColorMap.forEach { (schemeName, persistentMarkColor) ->
            markColorMap[schemeName] = persistentMarkColor.toMarkColor()
        }
        return markColorMap
    }

    private fun convertMarkColorMapToPersistentMarkColorMap(markColorMap: LinkedHashMap<String, MarkColor>): LinkedHashMap<String, PersistentMarkColor> {
        val persistentMarkColorMap = LinkedHashMap<String, PersistentMarkColor>()
        markColorMap.forEach { (schemeName, markColor) ->
            persistentMarkColorMap[schemeName] = PersistentMarkColor.of(markColor)
        }
        return persistentMarkColorMap
    }

    private fun initUI() {
        val persistentState = GlobalSettings.getPersistentState()
        markColorMap = convertPersistentMarkColorMapToMarkColorMap(persistentState.markColors)
        schemeInUse = persistentState.schemeName
        autoDetectConflictsWhenFileOpened = persistentState.autoDetectConflictsWhenFileOpened

        // make sure the built-in color is displayed at the top.
        cbColorScheme.addItem(BuiltInColor.DEFAULT_SCHEME_NAME)

        markColorMap.forEach { (schemeName, markColor) ->
            if (schemeName != BuiltInColor.DEFAULT_SCHEME_NAME) {
                cbColorScheme.addItem(schemeName)
            }
            allSchemeNames.add(schemeName)
        }
        cbAutoDetectConflictsWhenFileOpened.isSelected = autoDetectConflictsWhenFileOpened
        // Double check.
        if (schemeInUse in markColorMap) {
            cbColorScheme.selectedItem = schemeInUse
            markColorInUse.copyFrom(markColorMap[schemeInUse]!!)
        } else {
            cbColorScheme.selectedItem = BuiltInColor.DEFAULT_SCHEME_NAME
            markColorInUse.copyFrom(BuiltInColor.DEFAULT)
        }
        updateUI(markColorInUse)
    }

    private fun updateUI(color: MarkColor?) {
        if (color == null) {
            panelCustom.isVisible = false
            btnDelete.isEnabled = false
            return
        }
        if (color.isBuiltIn) {
            panelCustom.isVisible = false
            btnDelete.isEnabled = false
        } else {
            panelCustom.isVisible = true
            btnDelete.isEnabled = true
        }
        lbCurrentContent.text = color.currentContentBackground.formatHexA()
        lbCurrentHeader.text = color.currentHeaderBackground.formatHexA()
        lbIncomingContent.text = color.incomingContentBackground.formatHexA()
        lbIncomingHeader.text = color.incomingHeaderBackground.formatHexA()
        lbCommonContent.text = color.commonContentBackground.formatHexA()
        lbCommonHeader.text = color.commonHeaderBackground.formatHexA()
        lbCurrentContentBackground.background = color.currentContentBackground
        lbCurrentHeaderBackground.background = color.currentHeaderBackground
        lbIncomingContentBackground.background = color.incomingContentBackground
        lbIncomingHeaderBackground.background = color.incomingHeaderBackground
        lbCommonContentBackground.background = color.commonContentBackground
        lbCommonHeaderBackground.background = color.commonHeaderBackground
    }

    private fun setListeners() {
        lbCurrentContentBackground.addMouseListener(ColorChooserListener(CURRENT_CONTENT))
        lbCurrentHeaderBackground.addMouseListener(ColorChooserListener(CURRENT_HEADER))
        lbIncomingContentBackground.addMouseListener(ColorChooserListener(INCOMING_CONTENT))
        lbIncomingHeaderBackground.addMouseListener(ColorChooserListener(INCOMING_HEADER))
        lbCommonContentBackground.addMouseListener(ColorChooserListener(COMMON_CONTENT))
        lbCommonHeaderBackground.addMouseListener(ColorChooserListener(COMMON_HEADER))

        // Call when the selected item is changed.
        cbColorScheme.addItemListener { e: ItemEvent ->
            if (e.stateChange == ItemEvent.SELECTED) {
                val selectedItem = e.item as String
                updateUI(markColorMap[selectedItem])
            }
        }
        // New color scheme
        btnNew.addActionListener { e: ActionEvent? ->
            val newSchemeName = JOptionPane.showInputDialog(mainPanel, "Please input new scheme name: ")
                ?: return@addActionListener
            if ("" == newSchemeName) {
                JOptionPane.showMessageDialog(
                    mainPanel, "Scheme name can not be empty!", "Failed",
                    JOptionPane.ERROR_MESSAGE
                )
                return@addActionListener
            }
            if (allSchemeNames.contains(newSchemeName)) {
                JOptionPane.showMessageDialog(
                    mainPanel, "Scheme name already exists!", "Failed",
                    JOptionPane.ERROR_MESSAGE
                )
                return@addActionListener
            }
            lbCurrentContent.text = BuiltInColor.DEFAULT.currentContentBackground.formatHexA()
            lbCurrentHeader.text = BuiltInColor.DEFAULT.currentHeaderBackground.formatHexA()
            lbIncomingContent.text = BuiltInColor.DEFAULT.incomingContentBackground.formatHexA()
            lbIncomingHeader.text = BuiltInColor.DEFAULT.incomingHeaderBackground.formatHexA()
            lbCommonContent.text = BuiltInColor.DEFAULT.commonContentBackground.formatHexA()
            lbCommonHeader.text = BuiltInColor.DEFAULT.commonHeaderBackground.formatHexA()
            lbCurrentContentBackground.background = BuiltInColor.DEFAULT.currentContentBackground
            lbCurrentHeaderBackground.background = BuiltInColor.DEFAULT.currentHeaderBackground
            lbIncomingContentBackground.background = BuiltInColor.DEFAULT.incomingContentBackground
            lbIncomingHeaderBackground.background = BuiltInColor.DEFAULT.incomingHeaderBackground
            lbCommonContentBackground.background = BuiltInColor.DEFAULT.commonContentBackground
            lbCommonHeaderBackground.background = BuiltInColor.DEFAULT.commonHeaderBackground
            allSchemeNames.add(newSchemeName)
            cbColorScheme.addItem(newSchemeName)
            cbColorScheme.selectedItem = newSchemeName
            // add to markColor map
            markColorMap[newSchemeName] = MarkColor(newSchemeName, false)
            updateUI(markColorMap[newSchemeName])
        }
        // delete a color scheme
        btnDelete.addActionListener { e: ActionEvent? ->
            val selectedItem = cbColorScheme.selectedItem as String
            val confirm = JOptionPane.showConfirmDialog(
                mainPanel,
                "Delete scheme: $selectedItem ?",
                "Delete",
                JOptionPane.OK_CANCEL_OPTION
            )
            if (confirm == JOptionPane.OK_OPTION) {
                val markColor = markColorMap[selectedItem]!!
                // Double check. Prevent the built-in color scheme from being deleted.
                if (!markColor.isBuiltIn) {
                    markColorMap.remove(selectedItem)
                    cbColorScheme.removeItem(selectedItem)
                    allSchemeNames.remove(selectedItem)
                } else {
                    JOptionPane.showMessageDialog(
                        mainPanel, "The built-in color scheme cannot be deleted.",
                        "Failed", JOptionPane.ERROR_MESSAGE
                    )
                }
            }
        }
    }

    private inner class ColorChooserListener(private val fieldIndex: Int) : MouseListener {

        override fun mouseClicked(e: MouseEvent) {
            val selectedItem = cbColorScheme.selectedItem as String
            val tempColor = markColorMap[selectedItem]!!
            val initialColor: Color? = when (fieldIndex) {
                CURRENT_CONTENT -> {
                    tempColor.currentContentBackground
                }

                CURRENT_HEADER -> {
                    tempColor.currentHeaderBackground
                }

                INCOMING_CONTENT -> {
                    tempColor.incomingContentBackground
                }

                INCOMING_HEADER -> {
                    tempColor.incomingHeaderBackground
                }

                COMMON_CONTENT -> {
                    tempColor.commonContentBackground
                }

                COMMON_HEADER -> {
                    tempColor.commonHeaderBackground
                }

                else -> null
            }
            val chooseColor = ColorChooserService.getInstance().showDialog(
                null, mainPanel, "Choose a Color", initialColor, true,
                emptyList(), false
            )
            if (chooseColor != null) {
                when (fieldIndex) {
                    CURRENT_CONTENT -> {
                        lbCurrentContentBackground.background = chooseColor
                        lbCurrentContent.text = chooseColor.formatHexA()
                        tempColor.currentContentBackground = chooseColor
                    }

                    CURRENT_HEADER -> {
                        lbCurrentHeaderBackground.background = chooseColor
                        lbCurrentHeader.text = chooseColor.formatHexA()
                        tempColor.currentHeaderBackground = chooseColor
                    }

                    INCOMING_CONTENT -> {
                        lbIncomingContentBackground.background = chooseColor
                        lbIncomingContent.text = chooseColor.formatHexA()
                        tempColor.incomingContentBackground = chooseColor
                    }

                    INCOMING_HEADER -> {
                        lbIncomingHeaderBackground.background = chooseColor
                        lbIncomingHeader.text = chooseColor.formatHexA()
                        tempColor.incomingHeaderBackground = chooseColor
                    }

                    COMMON_CONTENT -> {
                        lbCommonContentBackground.background = chooseColor
                        lbCommonContent.text = chooseColor.formatHexA()
                        tempColor.commonContentBackground = chooseColor
                    }

                    COMMON_HEADER -> {
                        lbCommonHeaderBackground.background = chooseColor
                        lbCommonHeader.text = chooseColor.formatHexA()
                        tempColor.commonHeaderBackground = chooseColor
                    }
                }
                markColorMap[selectedItem] = tempColor
            }
        }

        override fun mousePressed(e: MouseEvent) {}
        override fun mouseReleased(e: MouseEvent) {}
        override fun mouseEntered(e: MouseEvent) {}
        override fun mouseExited(e: MouseEvent) {}
    }

    override fun getId(): String {
        return Constants.Resource.SETTINGS_TITLE
    }

    override fun getDisplayName(): String {
        return this.id
    }

    override fun createComponent(): JComponent? {
        return mainPanel
    }

    private fun isColorSchemeModified(): Boolean {
        if (cbColorScheme.selectedItem !== schemeInUse) {
            return true
        }
        if (markColorInUse != markColorMap[cbColorScheme.selectedItem as String]) {
            return true
        }
        return false
    }

    override fun isModified(): Boolean {
        if (isColorSchemeModified()) {
            return true
        }
        if (cbAutoDetectConflictsWhenFileOpened.isSelected != autoDetectConflictsWhenFileOpened) {
            return true
        }
        return false
    }

    override fun reset() {
        cbColorScheme.selectedItem = schemeInUse
        cbAutoDetectConflictsWhenFileOpened.isSelected = autoDetectConflictsWhenFileOpened
    }

    override fun apply() {
        val isColorSchemeModified = isColorSchemeModified()
        val isModified = isModified()
        val persistentState = GlobalSettings.getPersistentState()
        persistentState.markColors = convertMarkColorMapToPersistentMarkColorMap(markColorMap)
        persistentState.schemeName = (cbColorScheme.selectedItem as String)
        persistentState.autoDetectConflictsWhenFileOpened = cbAutoDetectConflictsWhenFileOpened.isSelected
        if (isModified) {
            schemeInUse = cbColorScheme.selectedItem as String
            markColorInUse.copyFrom(markColorMap[schemeInUse]!!)
            autoDetectConflictsWhenFileOpened = cbAutoDetectConflictsWhenFileOpened.isSelected
            Global.currentColor = markColorInUse
            Global.autoDetectConflictsWhenFileOpened = autoDetectConflictsWhenFileOpened
            if (isColorSchemeModified) {
                // println("[ColorSchemeModified] Global.currentColor: ${Global.currentColor}")
                HighlightConflictAction.refreshHighlighters()
            }
        }
    }
}
